zerver/tests/test_reactions.py

```diff
@@ -479,6 +479,136 @@ def test_remove_event(self) -> None:
         self.assertEqual(event["emoji_name"], "smile")
         self.assertEqual(event["message_id"], pm_id)
 
+    def test_reaction_event_scope(self) -> None:
+        iago = self.example_user("iago")
+        hamlet = self.example_user("hamlet")
+        polonius = self.example_user("polonius")
+        reaction_info = {
+            "emoji_name": "smile",
+        }
+
+        # Test `invite_only` streams with `!history_public_to_subscribers` and `!is_web_public`
+        stream = self.make_stream(
+            "test_reactions_stream", invite_only=True, history_public_to_subscribers=False
+        )
+        self.subscribe(iago, stream.name)
+        message_before_id = self.send_stream_message(
+            iago, "test_reactions_stream", "before subscription history private"
+        )
+        # Hamlet and Polonius joined after the message was send, hence must not
+        # recieve events of the messages sent before they joined the stream
+        self.subscribe(hamlet, stream.name)
+        self.subscribe(polonius, stream.name)
+        # No event expected by subscribers joining after message sent
+        # event only for reaction sender
+        events: List[Mapping[str, Any]] = []
+        with tornado_redirected_to_list(events):
+            result = self.api_post(
+                iago, f"/api/v1/messages/{message_before_id}/reactions", reaction_info
+            )
+        self.assert_json_success(result)
+        event = events[0]["event"]
+        self.assertEqual(event["type"], "reaction")
+        event_user_ids = set(events[0]["users"])
+        self.assertEqual(event_user_ids, {iago.id})
+        remove = self.api_delete(
+            iago, f"/api/v1/messages/{message_before_id}/reactions", reaction_info
+        )
+        self.assert_json_success(remove)
+
+        # Reaction for Message sent after subscription, should trigger events for Iago, Hamlet and Polonius
```

Comment:
Hi :) As the author of the previous PR, I'm happy to see this evolve and to learn something (I'm a new contributor).
I have a small question regarding this part of the tests: Shouldn't we also explicitly test the case here that Hamlet and Polonius joined after the message has been sent and now are able to still receive the event? (This would also apply to the next two test cases...)

Reply:
The tests are written keeping all the edges in mind. `message_before_id` is the `id` of the message send before they joined and `message_after_id` is the `id` of the message send after they joined. All the tests are based on that.

For streams the following tests are implemented:
1. Line 505: `iago` added reaction to `message_before` and none of `polonious` or `hamlet` got the event because the history is private.
2. Line 524: `iago` added reaction to `message_after` and everyone got the event because the messages were sent after they joined.
3. Line 543: `iago` added reaction to `message_before` and `hamlet` got the event but not `polonious` because he is a guest and we don't provide the history to subscribed guests if the stream is `!not_web_public` even after history is made public for subscribers.
4. Line 562: `iago` added reaction to `message_before` and everyone subscribed to the event because history is public and stream `is_web_public`.

All the above are written assuming that the sender is always notified of his message event.

I think this covers all cases. If you still feel anything is missed out feel free to reach out or reply to this message.

Reply:
The case mentioned by you is case 4.

Reply:
Ah, thank you very much for the explanation, @LoopThrough-i-j! :+1: Now I see it.